graphprot/NeuralNet.py

```diff
@@ -53,61 +55,142 @@ def __init__(self, database, Net,
+        self.class_weights = class_weights
+
+        # put the model
+        if self.task == 'reg' :
+            self.model = Net(dataset.get(0).num_features).to(self.device)
+            
+        elif self.task == 'class' :
+            self.classes = classes
+            self.classes_idx = {i: idx for idx, i in enumerate(self.classes)}
+            self.output_shape = len(self.classes)
+            try :
+                self.model = Net(dataset.get(0).num_features, self.output_shape).to(self.device)
+            except :
+                raise ValueError(
+                    f"The loaded model does not accept output_shape = {self.output_shape} argument \n\t"
+                    f"Check your input or adapt the model\n\t"
+                    f"Example :\n\t"
+                    f"def __init__(self, input_shape): --> def __init__(self, input_shape, output_shape) \n\t"
+                    f"self.fc2 = torch.nn.Linear(64, 1) --> self.fc2 = torch.nn.Linear(64, output_shape) \n\t")
+        
+        # optimizer
+        self.optimizer = torch.optim.Adam(
+            self.model.parameters(), lr=0.01)
+        
+        # loss 
+        if self.task == 'reg':
+            self.loss = MSELoss()
+
+        elif self.task == 'class':
+            self.loss = nn.CrossEntropyLoss(weight = self.class_weights, reduction='mean')
+
+        else:
+            raise ValueError(
+                f"Task {self.task} not recognized. Options are:\n\t "
+                f"reg': regression \n\t 'class': classifiation\n")        
+
 
     def train(self, nepoch=1, validate=False):
 
         self.model.train()
         for epoch in range(1, nepoch+1):
             t0 = time()
-            loss = self._epoch(epoch)
+            acc, loss = self._epoch(epoch)
 
             if validate:
-                _, val_loss = self.eval(self.valid_loader)
+                _, val_acc, val_loss = self.eval(self.valid_loader)
                 t = time() - t0
-                print('Epoch [%04d] : train loss %e | valid loss %e | time %1.2e sec.' % (
-                    epoch, loss, val_loss, t))
+                if acc is not None :
+                    print('Epoch [%04d] : train loss %e | train accuracy %1.4e | valid loss %e | val accuracy %1.4e | time %1.2e sec.' % (
+                        epoch, loss, acc, val_loss, val_acc, t))
+                else :
+                    print('Epoch [%04d] : train loss %e | valid loss %e | time %1.2e sec.' % (
+                        epoch, loss, val_loss, t))
+
             else:
                 t = time() - t0
-                print('Epoch [%04d] : train loss %e | time %1.2e sec.' % (
+                if acc is not None :
+                    print('Epoch [%04d] : train loss %e | accuracy %1.4e | time %1.2e sec. |' % (
+                    epoch, loss, acc, t))
+                else :
+                    print('Epoch [%04d] : train loss %e | time %1.2e sec. |' % (
                     epoch, loss, t))
```

Comment:
I think it would be more elegan to have : 
```python
acc, loss = self._epoch()
print('Epoch [%04d] : train loss | accuracy .... ')
if validate:
    valid_acc, valid_train = self.eval()
    print('Epoch [%04d] : valid loss valid acc')
```
cause now you have a lot of code duplication which is never ideal :)